pkg/controllers/provisioning/v1alpha1/reallocation/terminate.go

```diff
@@ -19,18 +19,70 @@ import (
 	"fmt"
 	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha1"
 	"github.com/awslabs/karpenter/pkg/cloudprovider"
+	"github.com/awslabs/karpenter/pkg/utils/ptr"
+	"github.com/awslabs/karpenter/pkg/utils/scheduling"
 	"go.uber.org/zap"
 	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 type Terminator struct {
 	kubeClient    client.Client
 	cloudprovider cloudprovider.Factory
+	coreV1Client  corev1.CoreV1Interface
+}
+
+// DrainNodes takes in a list of expired non-drained nodes and calls drain on them
```

Comment:
For the refactor, Ive decided to call delete in the drain nodes action. For any node of the iteration, I will only log a debug statement if it fails and continues. Let me know what you think.

Reply:
Added some comments. I think it's pretty rare that you want to continue if these API calls fail. In general, only continue if it would be harmful not to. The main idea behind this is that if the API server is rejecting your request, it probably doesn't help to keep spamming it with more requests. If it's overloaded, it's bad, if it's an authentication issue, it won't help, if it's unable to connect, you're probably stuck anyways.

Controller runtime has some great machinery around tracking reconcile loop failures and plugging them into a backoff/retry function. 

There are some cases where you see "continuing after ...". These are mostly useful because we've already created an EC2 instance, and it would be harmful to not attempt to at least bind the pods we prepared for it. I'm also going to deep dive all of those continue statements and make sure they're all good ideas under the framework mentioned about.

Reply:
Got it. The EC2 terminate instance call can take in a batch ids so I don't think we should be concerned with this, right?

Reply:
Probably; I would test to see what happens if you pass in a batch but have (say) one of the instances marked not terminatable.